gravity benchmark thin shell (#4164)
geodynamics/aspect: doc/manual/manual.tex

The parameter file and initial temperature model for this benchmark can be found at \url{benchmarks/yamauchi_takei_2016_anelasticity/yamauchi_takei_2016_anelasticity.prm} and \url{benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc}. Code to recreate  Fig.~\ref{fig:anelasticity}  is provided in \url{benchmarks/yamauchi_takei_2016_anelasticity/plot_output}.

\subsubsection{Thin shell gravity benchmark}
\label{sec:benchmark-thin-shell-gravity}

> \textit{This section was contributed by Cedric Thieulot, Bart Root and Paul Bremmer.}

Reviewer: ```suggestion
\textit{This section was contributed by Cedric Thieulot, Bart Root and Paul Bremner.}
```

Author: oups!!!

> The domain is a spherical shell of 10~\si{\km} radius centered at a depth $D$, i.e. the inner radius is $R_i=6371-D-5~\si{\km}$ and the outer radius $R_o=6371-D+5~ \si{\km}$. It is filled with a fluid of constant density $\rho_0=3300~\si{\kg\per\cubic\metre}$ with total mass $M=\frac43 \pi (R_o^3-R_i^3)$.

Reviewer: In the manual, subscripts that aren't indices are typically in roman typeface. The easiest way to do this is, for example: `$R\textsubscript{inner}=6371-D-5~\si{\km}$`. Same for `outer`, rather than `$o$`

Reviewer 2: Separately, would you mind adding a short paragraph at the top of this section that starts it out by saying what the goal here is? You're jumping right into the weeds :-)

Author: did that

Author: also: fun fact, M is wrong here :)

> \noindent Both the Stokes and energy equations solve are bypassed:

Reviewer 2: Leave formatting to latex :-)
```suggestion
Both the Stokes and energy equations solve are bypassed:
```

\lstinputlisting[language=prmfile]{cookbooks/benchmarks/gravity_thin_shell/doc/thinshell_b.prm.out}

> We make use of the Custom mesh subdivision option to generate a mesh where a single cell is used in the radial direction (parameterized with the number of 'slices') while 6 blocks (default value) of $2^5\times 2^5$ cells are used in the lateral direction. This gives a total of 6,144 mesh cells. For $D=100~\si{\km}$ the parameterization of the mesh is then as follows:

Reviewer 2: ```suggestion
We make use of the Custom mesh subdivision option to generate a mesh where a single cell is used in the radial direction (parameterized with the number of `slices') while 6 blocks (default value) of $2^5\times 2^5$ cells are used in the lateral direction. This gives a total of 6,144 mesh cells. For $D=100~\si{\km}$ the parameterization of the mesh is then as follows:
```

\lstinputlisting[language=prmfile]{cookbooks/benchmarks/gravity_thin_shell/doc/thinshell_a.prm.out}

> We wish to compute the gravity potential $U$ and the gravity vector ${\mathbf g}$ at a height of $250~\si{\km}$ above the surface of the Earth (i.e. the satellite is flying at a radius of $R_s=6371+250=6621~\si{\km}$) on a regular longitude-latitude grid spanning the whole Earth, with a $2\si{\degree}\times 2\si{\degree}$ resolution:

Reviewer: ```suggestion
We wish to compute the gravity potential $U$ and the gravity vector ${\mathbf g}$ at a radius of $R\textsubscript{s}=6371+250=6621~\si{\km}$ (i.e. the gravity experienced by a satellite flying at a height of $250~\si{\km}$ above the surface of the Earth) on a regular longitude-latitude grid spanning the whole Earth, with a $2\si{\degree}\times 2\si{\degree}$ resolution:
```

Author: done

\lstinputlisting[language=prmfile]{cookbooks/benchmarks/gravity_thin_shell/doc/thinshell_c.prm.out}

> The Gauss-Legendre Quadrature (GLQ) algorithm is central to the ASPECT code as it is used to compute the elemental integrals stemming from the discretisation of the weak form of the PDEs. Logically, the gravity postprocessor is also based on the GLQ since the potential $U$ at any location in space is given by the integral 

Reviewer: ```suggestion
The Gauss-Legendre Quadrature (GLQ) algorithm is central to the ASPECT code as it is used to compute the elemental integrals stemming from the discretization of the weak form of the PDEs. Logically, the gravity postprocessor is also based on the GLQ since the potential $U$ at any location in space is given by the integral 
```

Reviewer 2: ```suggestion
The Gauss-Legendre Quadrature (GLQ) algorithm is central to the \aspect{} code as it is used to compute the elemental integrals stemming from the discretisation of the weak form of the PDEs. Logically, the gravity postprocessor is also based on the GLQ since the potential $U$ at any location in space is given by the integral 
```

Reviewer 2: @bobmyhill That must have hurt to correct @cedrict to use AE spelling :-)

Author: exactly my thought :)

\begin{equation}
U({\mathbf r}) = \iiint_\Omega \frac{G \rho({\mathbf r}')}{|{\mathbf r}-{\mathbf r}'|} d{\mathbf r}'
= \sum_K \iiint_{\Omega_K} \frac{G \rho({\mathbf r}')}{|{\mathbf r}-{\mathbf r}'|} d{\mathbf r}'
\end{equation}

> where the sum runs over all cells of the mesh and ${G}=6.67430\times 10^{-11}~\si{\cubic\metre\per\kg\per\square\second}$ is the gravitaional constant. The gravity acceleration vector is obtained via ${\mathbf g}=-{\mathbf \nabla} U$, or 

Reviewer: ```suggestion
where the sum runs over all cells of the mesh and ${G}=6.67430\times 10^{-11}~\si{\cubic\metre\per\kg\per\square\second}$ is the gravitational constant. The gravity acceleration vector is obtained via ${\mathbf g}=-{\mathbf \nabla} U$, or 
```

\begin{equation}
{\mathbf g}({\mathbf r}) = 
\iiint_\Omega  \frac{G \rho({\mathbf r}')}{|{\mathbf r}-{\mathbf r}'|^2} d{\mathbf r}'
= 
\sum_K \iiint_{\Omega_K} \frac{G \rho({\mathbf r}')}{|{\mathbf r}-{\mathbf r}'|^2} d{\mathbf r}'.
\end{equation}

> The default number of quadrature points for each cell in the postprocessor is $2^{ndim}$, where $ndim$ is the number of dimensions. The $n$-point GLQ allows exact integration of $(2n-1)$-order polynomials. However, the integrand of the Newton integral is not a polynomial (it contains a term $\sim r^{-m}$), so there is not an optimal number of quadrature points to use. Therefore, the postprocessor allows the user to choose how many additional points per dimension are used with an expectation that an increase in the number of quadrature points inside the cells leads to a more accurate calculation. This increase number $I$ is set to 1 in the example above, although it can also be chosen to be 0, 1, 2 or even -1. The case $I=-1$ approximates the cell as a point mass since there is a single quadrature point in the middle of the cell.

Reviewer: ```suggestion
The default number of quadrature points for each cell in the postprocessor is $2\textsuperscript{ndim}$, where ndim is the number of dimensions. The $n$-point GLQ allows exact integration of $(2n-1)$-order polynomials. However, the integrand of the Newton integral is not a polynomial (it contains a term $\sim r^{-m}$), so there is not an optimal number of quadrature points to use. Therefore, the postprocessor allows the user to choose how many additional points per dimension are used with an expectation that an increase in the number of quadrature points inside the cells leads to a more accurate calculation. This increase number $I$ is set to 1 in the example above, although it can also be chosen to be 0, 1, 2 or even -1. The case $I=-1$ approximates the cell as a point mass since there is a single quadrature point in the middle of the cell.
```

Author: done

> Given the symmetry of the problem the value of the potential and acceleration depend solely on the distance $r$ from the origin, see Turcotte and Schubert \cite{TS14}. Their analytical values are given by $U(R_s) = - {\cal G}M/R_s$ and $|{\mathbf g}(R_s)|= g_r(R_s)={\cal G}M/R_s^2$. 

Reviewer: more opportunity for \textsubscript{s}.

Author: indeed! exciting! :)

-144088.7937 &
-144088.7937 \\
& a.v. && 2176.2391 &&& -144088.7937 & &\\
\hline\hline
1500&-1 & $1^3$ & 
1498.7997 &
1498.7569 &
1499.0522 &
-99236.0403 &
-99238.4122 &
-99235.3167 \\
& 0 & $2^3$ &
1498.0239 &
1498.0236 &
1498.0240 &
-99184.1673 &
-99184.1678 &
-99184.1647 \\
& 1 & $3^3$ &
1498.0240 &
1498.0240 & 
1498.0240 &
-99184.1672 &
-99184.1672 &
-99184.1672 \\
& 2 & $4^3$ &
1498.0240 &
1498.0240 &
1498.0240 &
-99184.1672 &
-99184.1672 &
-99184.1672 \\
& a.v. && 1498.0240 &&& -99184.1672 &&\\
\hline\hline
3000 &-1 & $1^3$ &   
717.8387 &
717.8315 & 
717.8533 &
-47528.1777 &
-47528.3488 &
-47528.0725 \\
 & 0 & $2^3$ &   
717.4641 &
717.4640 &
717.4641 &
-47503.2981 &
-47503.2982 &
-47503.2980 \\
 & 1 & $3^3$ &   
717.4641 &
717.4641 &
717.4641 &
-47503.2981 &
-47503.2981 &
-47503.2981 \\
& 2 & $4^3$ & 
717.4641 &
717.4641 &
717.4641 &
-47503.2981 &
-47503.2981 &
-47503.2981 \\
&{\small a.v.} && 717.4641 &&& -47503.2981 && \\
\hline
\end{tabular}\\
\caption{\it Thin shell gravity benchmark: $1\si{mGal}=10^{-5}\si{\metre\per\square\second}$. $n_q$ is the number of GLQ points per cell. 'a.v.' stands for analytical value.}
\label{tab:thin_shell_gravity_benchmark}
\end{table}

> The accuracy of the calculations increases with $I$ but so does the time spent in the postprocessor: for $I\in\{-1,0,1,2\}$  this time was about 18, 132, 440 and 1040~\si{s} respectively. This is easily explained when one realises that increasing $I$ from 0 to 1 means that the number of GLQ points per cell goes from $2^3=8$ to $3^3=27$, i.e. a $3.375$ increase in operations for the same number of cells and measurement points. Coincidentally, the time spent in the postprocessor increases by the same amount $440/132\sim 3.4$.

Reviewer: ```suggestion
The accuracy of the calculations increases with $I$ but so does the time spent in the postprocessor: for $I\in\{-1,0,1,2\}$  this time was about 18, 132, 440 and 1040~\si{s} respectively. This is easily explained when one realises that increasing $I$ from 0 to 1 means that the number of GLQ points per cell goes from $2^3=8$ to $3^3=27$, i.e. a $3.375$ increase in operations for the same number of cells and measurement points. The time spent in the postprocessor increases by a similar factor ($440/132\sim 3.4$).
```

Reviewer: You could also quantify the relative accuracies when using the different values of $I$ here.

Reviewer 2: ```suggestion
The accuracy of the calculations increases with $I$ but so does the time spent in the postprocessor: for $I\in\{-1,0,1,2\}$  this time was about 18, 132, 440 and 1040~\si{s} respectively. This is easily explained when one realizes that increasing $I$ from 0 to 1 means that the number of GLQ points per cell goes from $2^3=8$ to $3^3=27$, i.e. a $3.375$ increase in operations for the same number of cells and measurement points. Coincidentally, the time spent in the postprocessor increases by the same amount $440/132\sim 3.4$.
```

Author: this benchmark is part of a paper about to be submitted in which we discuss the errors and compare those with errors obtained with other codes/methods.

> The accuracy obtained with lower $I$ values increases with increasing anomaly depth (or increasing distance from the observation point). Note that this simplistic case has uniform density so that the projection of the density from the nodes onto the quadrature points is exact and it does not introduce any smoothing of data which might occur in the presence of density discontinuities (e.g. air-water, water-crust, moho, etc.) inside a cell.

Reviewer: ```suggestion
The accuracy obtained with lower $I$ values increases with increasing anomaly depth (or increasing distance from the observation point). Note that this benchmark has uniform density so that the projection of the density from the nodes onto the quadrature points is exact and it does not introduce any smoothing of data which might occur in the presence of density discontinuities (e.g. air-water, water-crust, moho, etc.) inside a cell.
```

> In conclusion, $I=1$ seems to be the best compromise between accuracy and compute time for this experiment.

Reviewer: ```suggestion
$I=1$ seems to be the best compromise between accuracy and compute time for this experiment.
```

Reviewer: It might be worth another sentence explaining how you define "best". What kind of accuracy is desirable?

Author: done

The parameter file and initial temperature model for this benchmark can be found at \url{benchmarks/yamauchi_takei_2016_anelasticity/yamauchi_takei_2016_anelasticity.prm} and \url{benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc}. Code to recreate  Fig.~\ref{fig:anelasticity}  is provided in \url{benchmarks/yamauchi_takei_2016_anelasticity/plot_output}.

\subsubsection{Thin shell gravity benchmark}
\label{sec:benchmark-thin-shell-gravity}

\textit{This section was contributed by Cedric Thieulot, Bart Root and Paul Bremner.}

> The idea behind this benchmark is to test the accuracy of the gravity postprocessor: the domain is a thin shell of constant density somewhere in the Earth mantle and we wish to compute the resultiong gravity field and potential at satellite height.

Reviewer: ```suggestion
The idea behind this benchmark is to test the accuracy of the gravity postprocessor: the domain is a thin shell of constant density somewhere in the Earth mantle and we wish to compute the resulting gravity field and potential at satellite height.
```

-144088.7937 &
-144088.7937 \\
& a.v. && 2176.2391 &&& -144088.7937 & &\\
\hline\hline
1500&-1 & $1^3$ & 
1498.7997 &
1498.7569 &
1499.0522 &
-99236.0403 &
-99238.4122 &
-99235.3167 \\
& 0 & $2^3$ &
1498.0239 &
1498.0236 &
1498.0240 &
-99184.1673 &
-99184.1678 &
-99184.1647 \\
& 1 & $3^3$ &
1498.0240 &
1498.0240 & 
1498.0240 &
-99184.1672 &
-99184.1672 &
-99184.1672 \\
& 2 & $4^3$ &
1498.0240 &
1498.0240 &
1498.0240 &
-99184.1672 &
-99184.1672 &
-99184.1672 \\
& a.v. && 1498.0240 &&& -99184.1672 &&\\
\hline\hline
3000 &-1 & $1^3$ &   
717.8387 &
717.8315 & 
717.8533 &
-47528.1777 &
-47528.3488 &
-47528.0725 \\
 & 0 & $2^3$ &   
717.4641 &
717.4640 &
717.4641 &
-47503.2981 &
-47503.2982 &
-47503.2980 \\
 & 1 & $3^3$ &   
717.4641 &
717.4641 &
717.4641 &
-47503.2981 &
-47503.2981 &
-47503.2981 \\
& 2 & $4^3$ & 
717.4641 &
717.4641 &
717.4641 &
-47503.2981 &
-47503.2981 &
-47503.2981 \\
&{\small a.v.} && 717.4641 &&& -47503.2981 && \\
\hline
\end{tabular}\\
\caption{\it Thin shell gravity benchmark: $1\si{mGal}=10^{-5}\si{\metre\per\square\second}$. $n_q$ is the number of GLQ points per cell. 'a.v.' stands for analytical value.}
\label{tab:thin_shell_gravity_benchmark}
\end{table}

> The accuracy of the calculations increases with $I$ but so does the time spent in the postprocessor: for $I\in\{-1,0,1,2\}$  this time was about 18, 132, 440 and 1040~\si{s} respectively. This is easily explained when one realizes that increasing $I$ from 0 to 1 means that the number of GLQ points per cell goes from $2^3=8$ to $3^3=27$, i.e. a $3.375$ increase in operations for the same number of cells and measurement points. The time spent in the postprocessor increases by a similar factor $440/132\sim 3.4$.

Reviewer: ```suggestion
The accuracy of the calculations increases with $I$ but so does the time spent in the postprocessor: for $I\in\{-1,0,1,2\}$  this time was about 18, 132, 440 and 1040~\si{s} respectively. This is easily explained when one realizes that increasing $I$ from 0 to 1 means that the number of GLQ points per cell goes from $2^3=8$ to $3^3=27$, i.e. a $3.375$ increase in operations for the same number of cells and measurement points. The time spent in the postprocessor increases by a similar factor ($440/132\sim 3.4$).
```

Author: done
